### ISSUE TITLE
Bypassing dynamic array limitation in JAX

### ISSUE THREAD

**Issue by unalmis** [flatironinstitute/jax-finufft#199]
Due to [limitations in JAX with dynamic array shape sizes](https://github.com/jax-ml/jax/discussions/24410), to interpolate a Fourier series to $Q$ non-uniform points with a type 2 nufft, one is often forced to pad the array `Q` to size `S` and instead interpolate to $|S| \sim |Q|^2$ points. So the cost becomes $O(K^d log(K^d) + S \log^d \epsilon^{-1})$ for the forward and similiar for the reverse autodiff gradient which further calls type 1 nuffts.

Given a boolean mask which denotes which points in the input array `S` are junk, would it be simple to add an option to finufft to skip the spreading computation to the junk points and just return some junk value at those indices? In generic python this would be as simple as [this](https://github.com/flatironinstitute/finufft/discussions/783#discussioncomment-15480295).

If this feature happens to be trivial to add, but would be considered bloat, I would be interested in adding this feature to a forked repository in the future, so it would still be useful to me to know that modification.

_Originally posted by @unalmis in https://github.com/flatironinstitute/finufft/discussions/783_

**Comment by unalmis** [comment]
As I understand, we should always be able to bypass some limitation to compute a function `f` in JAX by implementing `f` in an external language, using FFI callback, and defining the derivative. Hence, it would be nice to have this feature since it is potentially a simple modification (10 lines).

**Comment by unalmis** [comment]
The only way to do enable dynamic array shapes in JAX is to write a C++ code with FFI. Right now `jax-finufft` is doing that to use finufft. However, because of JAX's limitation with dynamic array shapes as described above, the performance of `jax-finufft` in practice is much slower than it would be by calling `finufft` directly.

It would be nice to add the suggestion in this PR so that JAX users can use `jax-finufft` with comparable performance to using finufft.

https://docs.jax.dev/en/latest/jax-primitives.html#xla-compilation-rules
https://docs.jax.dev/en/latest/ffi.html
